scipy/signal/signaltools.py

```diff
@@ -2438,18 +2480,48 @@ def resample_poly(x, up, down, axis=0, window=('kaiser', 5.0)):
     n_post_pad = 0
     n_pre_remove = (half_len + n_pre_pad) // down
     # We should rarely need to do this given our filter lengths...
-    while _output_len(len(h) + n_pre_pad + n_post_pad, x.shape[axis],
+    while _output_len(len(h) + n_pre_pad + n_post_pad, n_in,
                       up, down) < n_out + n_pre_remove:
         n_post_pad += 1
     h = np.concatenate((np.zeros(n_pre_pad, dtype=h.dtype), h,
                         np.zeros(n_post_pad, dtype=h.dtype)))
     n_pre_remove_end = n_pre_remove + n_out
 
+    # Remove background depending on the border option
+    funcs = {'mean': np.mean, 'median': np.median,
+             'minimum': np.amin, 'maximum': np.amax}
+    if border == 'zero':
+        background_line = None
+    elif border in funcs:
+        background_line = [funcs[border](x, axis=axis), 0]
+    elif border == 'line':
+        background_line = [x.take(0, axis),
+                           (x.take(-1, axis) - x.take(0, axis))*n_in/(n_in-1)]
+    else:
+        raise ValueError('border must be line, maximum, mean, median, minimum or zero')
+
+    if background_line is not None:
+        rel_len = np.linspace(0.0, 1.0, n_in, endpoint=False)
+        background_in = np.stack(
+            [background_line[0] + background_line[1]*l for l in rel_len],
+            axis=axis)
+        x = x - background_in.astype(x.dtype)
+
     # filter then remove excess
     y = upfirdn(h, x, up, down, axis=axis)
     keep = [slice(None), ]*x.ndim
     keep[axis] = slice(n_pre_remove, n_pre_remove_end)
-    return y[tuple(keep)]
+    y_keep = y[tuple(keep)]
+
+    # Add background back
+    if background_line is not None:
+        rel_len = np.linspace(0.0, 1.0, n_out, endpoint=False)
+        background_out = np.stack(
+            [background_line[0] + background_line[1]*l for l in rel_len],
+            axis=axis)
```

Comment:
To avoid using a list comprehension + np.stack to build `background_in` and `background_out`, you could define a utility function like:
```Python

def _reshape_nd(x1d, ndim, axis):
    """Reshape x1d to size 1 along all axes in ``range(ndim)`` except for ``axis``."""
    shape = [1] * ndim
    shape[axis] = x1d.size
    return x1d.reshape(shape)
```
And then you can create e.g. `background_out` via:
```Python
background_out = _reshape_nd(background_line[0] + background_line[1] * rel_len,
                             y.ndim,
                             axis)
```

Reply:
Not sure I get what you mean. The output of _reshape_nd is one-dimensional, but background_in and background_out must have the same dimensions as x.

Note that background_line[0] and background_line[1] have the same dimension as x in all axes except for the axis to be resampled.

Reply:
But maybe something like this will work and be better:
`background_out = np.expand_dims(background_line[0], axis) + np.expand_dims(background_line[1], axis) * _reshape_nd(rel_len)`

Reply:
>  The output of _reshape_nd is one-dimensional, but background_in and background_out must have the same dimensions as x.

okay, yeah I had misread that.